core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -244,6 +244,28 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
       leaderId = members.keys.headOption
   }
 
+  def maybeElectNewJoinedLeader(): Option[MemberMetadata] = {
+    leaderId match {
+      case Some(currentLeaderId) =>
+        val oldLeader = get(currentLeaderId)
+        if (!oldLeader.isAwaitingJoin) {
+          members.find(_._2.isAwaitingJoin) match {
+            case Some(anyJoinedMember) => {
+              leaderId = Option(anyJoinedMember._1)
+              info(s"Group leader [member.id: ${oldLeader.memberId}, " +
+                s"group.instance.id: ${oldLeader.groupInstanceId}] failed to join " +
+                s"before rebalance timeout, while new leader $leaderId was elected.")
+              Some(anyJoinedMember._2)
+            }
+            case _ => Some(oldLeader)
```

Comment:
nit: maybe use `case None` since it is the only possibility.

Also, I think it might be worth having a log message somewhere indicating that the group cannot complete the rebalance because there is no leader joined.